modules/pipeline/providers/leaderworker/leader_listen_worker_task.go

```diff
@@ -168,10 +169,66 @@ func (p *provider) listWorkerTasks(ctx context.Context, workerID worker.ID) ([]w
 	}
 	var tasks []worker.LogicTask
 	for _, kv := range resp.Kvs {
-		logicTaskID := p.getWorkerTaskLogicIDFromIncomingKey(workerID, string(kv.Key))
+		logicTaskID := p.getWorkerLogicTaskIDFromIncomingKey(workerID, string(kv.Key))
 		logicTaskData := kv.Value
 		task := worker.NewLogicTask(logicTaskID, logicTaskData)
 		tasks = append(tasks, task)
 	}
 	return tasks, nil
 }
+
+func (p *provider) leaderListenTaskCanceling(ctx context.Context) {
+	prefix := p.makeEtcdLeaderLogicTaskCancelListenPrefix()
+	p.ListenPrefix(ctx, prefix,
+		func(ctx context.Context, event *clientv3.Event) {
+			// concurrent cancel
+			go func() {
+				key := string(event.Kv.Key)
+				logicTaskID := p.getLogicTaskIDFromLeaderCancelKey(key)
+				// check logic task
+				isHandling, workerID := p.IsTaskBeingProcessed(ctx, logicTaskID)
+				if !isHandling {
+					p.Log.Warnf("skip cancel logic task(not being processed), logicTaskID: %s", logicTaskID)
+					return
+				}
+				// do cancel
+				distributedCancelKey := p.makeEtcdWorkerLogicTaskCancelKey(workerID, logicTaskID)
+				for {
+					_, err := p.EtcdClient.Put(ctx, distributedCancelKey, "")
+					if err != nil {
+						p.Log.Errorf("failed to distributed cancel logic task(auto retry), workerID: %s, logicTaskID: %s, err: %v", workerID, logicTaskID, err)
+						time.Sleep(p.Cfg.Leader.RetryInterval)
+						continue
+					}
+					break
+				}
+			}()
+		},
+		nil,
+	)
```

Comment:
I think it's better to put the logic to get the `distributedCancelKey` into the loop as well, because task's `IsTaskBeingProcessed` may changed when sleep retryinterval

Reply:
adjusted. use `rutil.ContinueWorking` and wrap from `check logic task` to the end.